### PR TITLE
Fast adapter support

### PR DESCRIPTION
**Pull request by mamin506**
.

### CODE REVIEW

File: src/runtime_src/core/common/drv/fast_adapter.c [hunk 1]
```diff
@@ -191,7 +191,7 @@ int xrt_cu_fa_init(struct xrt_cu *xcu)
 		goto err;
 	}
 
-	/* TODO: Write 0 to MSWR resiger is a hack.
+	/* TODO: Write 0 to MSWR resiger is a hack for unknown hardware issue.
```
Comment (keryell) [inline src/runtime_src/core/common/drv/fast_adapter.c:194]: Is "resiger" is a hack for an unknown hardware device? :-)

File: src/runtime_src/core/common/drv/include/xrt_cu.h [hunk 2]
```diff
@@ -253,6 +253,9 @@ struct xrt_cu {
 	 * one for submit, one for complete
 	 */
 	struct task_struct	  *thread;
+	/* Good for debug */
+	u32			   sleep_cnt;
+	u32			   max_running;
```
Comment (keryell) [inline src/runtime_src/core/common/drv/include/xrt_cu.h:258]: Avoid mixing tab and spaces...

File: src/runtime_src/core/common/drv/fast_adapter.c (hunk 1 above)
Comment (mamin506) [reply src/runtime_src/core/common/drv/fast_adapter.c]: Thanks. Fixed

File: src/runtime_src/core/common/drv/include/xrt_cu.h (hunk 2 above)
Comment (mamin506) [reply src/runtime_src/core/common/drv/include/xrt_cu.h:258]: Thanks for the comment! But we use space after tab for alignment at many place. The checkpatch.pl allow tab followed by spaces.
No change at this time until we plan to clean up this coding style.

File: src/runtime_src/core/common/drv/xrt_cu.c [hunk 3]
```diff
@@ -446,6 +459,83 @@ int xrt_cu_cfg_update(struct xrt_cu *xcu, int intr)
 	return err;
 }
 
+/* TODO: maybe we should move below fast adapter specific functions to
+ * fast_adapter.c. Before we clearly know what to do for below cases,
+ *	single CU  to single plram
+ *	multi  CUs to single plram
+ *	single CU  to multi  plrams
```
Comment (maxzhen) [inline src/runtime_src/core/common/drv/xrt_cu.c:466]: what is not clear here?

**Review by maxzhen — COMMENTED** [review]
did we add sync bo support for p2p bo?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/p2p.c [hunk 4]
```diff
@@ -1087,6 +1087,43 @@ static int p2p_refresh_rbar(struct platform_device *pdev)
...
+	}
+
+	bank_off = p2p_bar_map(p2p, bank_addr, bank_size,
+		XOCL_P2P_CHUNK_SIZE);
+	if (bank_off < 0) {
+		ret = -ENOENT;
+		goto failed;
+	}
+
+	p2p_bar_unmap(p2p, bank_off);
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/p2p.c:1117]: The released map will be reused by next map request. And it looks the same with p2p_remap_resource()? Maybe you may change p2p_remap_resource() and call it? you do not have to release if you will used the address for the entire lifecycle of xclbin.

File: src/runtime_src/core/common/drv/xrt_cu.c (hunk 3 above)
Comment (mamin506) [reply src/runtime_src/core/common/drv/xrt_cu.c:466]: We discussed this before. Just forgot to delete. Will clean this comment.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/p2p.c (hunk 4 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/p2p.c:1117]: Oh.. let me investigate p2p_remap_resource(). Thanks for the comment.

**Comment by mamin506** [comment]
> did we add sync bo support for p2p bo?

Yes.

**Review by maxzhen — APPROVED** [review]

**Review by houlz0507 — APPROVED** [review]

**Comment by houlz0507** [comment]
This is kind of urgent. Merge it for now to unblock platform verification. And other review commends will be covered with following PRs

**Pull request merged**